### PR TITLE
for for all the watches to be called at least once before issuing a sync

### PR DESCRIPTION
**Pull request by yuval-k**
BOT NOTES: 
resolves https://github.com/solo-io/solo-kit/pull/222

### CODE REVIEW

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go [hunk 1]
```diff
@@ -191,7 +191,38 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
 		originalSnapshot := {{ .GoName }}Snapshot{}
 		currentSnapshot := originalSnapshot.Clone()
 		timer := time.NewTicker(time.Second * 1)
+
+
+		{{- range .Resources}}
+		{{- if not .ClusterScoped }}
+				{{ lower_camel .PluralName }}ByNamespace := make(map[string]{{ .ImportPrefix }}{{ .Name }}List)
```
Comment (yuval-k) [inline pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: @ilackarms is it safe to assume that watch never returns a nil list, even if it's empty?

Comment (ilackarms) [reply pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: generated clients return an empty list `var x FooList`
clients implementing the interface can return whatever they want

for what is the context?

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go [hunk 2]
```diff
@@ -126,13 +126,24 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
...
 {{- end}}
 
+	currentSnapshot := {{ .GoName }}Snapshot{}
+
 	for _, namespace := range watchNamespaces {
 {{- range .Resources}}
 {{- if (not .ClusterScoped) }}
 		/* Setup namespaced watch for {{ .Name }} */
+		{
+			upstreams, err := c.{{ lower_camel .Name }}.List(namespace, clients.ListOpts{Ctx: opts.Ctx, Selector: opts.Selector})
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: `upstreams` -> `{{ lower_camel .PluralName }}`

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go [hunk 3]
```diff
@@ -126,13 +126,24 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
...
+
 {{- end}}
 
+	currentSnapshot := {{ .GoName }}Snapshot{}
+
 	for _, namespace := range watchNamespaces {
 {{- range .Resources}}
 {{- if (not .ClusterScoped) }}
 		/* Setup namespaced watch for {{ .Name }} */
+		{
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: why use a closure here?

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go (hunk 2 above)
Comment (yuval-k) [reply pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: no need as they are in an internal scope

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go (hunk 3 above)
Comment (yuval-k) [reply pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: so that i can re-use the variable name upstreams

**Pull request merged**